dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,837 @@
+        return E_FAIL;
+    }
+    return GetItemDetails(pidl, iColumn, psd, NULL);
+}
+
+int CCabFolder::MapSCIDToColumn(const SHCOLUMNID &scid)
+{
+    for (UINT i = 0; i < _countof(g_Columns); ++i)
+    {
+        if (g_Columns[i].pkg && IsEqual(scid, *g_Columns[i].pkg, g_Columns[i].pki))
+            return i;
+    }
+    return -1;
+}
+
+IFACEMETHODIMP CCabFolder::MapColumnToSCID(UINT column, SHCOLUMNID *pscid)
+{
+    if (column < _countof(g_Columns) && g_Columns[column].pkg)
+    {
+        pscid->fmtid = *g_Columns[column].pkg;
+        pscid->pid = g_Columns[column].pki;
+        return S_OK;
+    }
+    return E_FAIL;
+}
+
+IFACEMETHODIMP CCabFolder::EnumObjects(HWND hwndOwner, DWORD dwFlags, LPENUMIDLIST *ppEnumIDList)
+{
+    CEnumIDList *p = CEnumIDList::CreateInstance();
+    *ppEnumIDList = static_cast<LPENUMIDLIST>(p);
+    return p ? p->Fill(m_CurDir, hwndOwner, dwFlags) : E_OUTOFMEMORY;
+}
+
+IFACEMETHODIMP CCabFolder::BindToObject(PCUIDLIST_RELATIVE pidl, LPBC pbcReserved, REFIID riid, LPVOID *ppvOut)
+{
+    UNIMPLEMENTED;
+    return E_NOTIMPL;
+}
+
+HRESULT CCabFolder::CompareID(LPARAM lParam, PCUITEMID_CHILD pidl1, PCUITEMID_CHILD pidl2)
+{
+    CABITEM *p1 = (CABITEM*)pidl1, *p2 = (CABITEM*)pidl2;
+    HRESULT hr = S_OK;
+    int ret = 0;
+
+    if (lParam & (SHCIDS_ALLFIELDS | SHCIDS_CANONICALONLY))
+    {
+        ret = lstrcmpiW(p1->Path, p2->Path);
+        if (ret && (lParam & SHCIDS_ALLFIELDS))
+        {
+            for (UINT i = 0; ret && SUCCEEDED(hr) && i < COLCOUNT; ++i)
+            {
+                hr = (i == COL_NAME) ? 0 : CompareID(i, pidl1, pidl2);
+                ret = (short)HRESULT_CODE(hr);
+            }
+        }
+    }
+    else
+    {
+        UINT col = lParam & SHCIDS_COLUMNMASK;
+        switch (col)
+        {
+            case COL_NAME:
+                ret = StrCmpLogicalW(p1->GetName(), p2->GetName());
+                break;
+
+            case COL_SIZE:
+                ret = p1->Size - p2->Size;
+                break;
+
+            case COL_MDATE:
+                ret = MAKELONG(p1->Time, p1->Date) - MAKELONG(p2->Time, p2->Date);
+                break;
+
+            default:
+            {
+                if (col < COLCOUNT)
+                {
+                    SHELLDETAILS sd1, sd2;
+                    sd2.str.uType = STRRET_CSTR;
+                    if (SUCCEEDED(hr = GetDetailsOf(pidl1, col, &sd1)) && SUCCEEDED(hr = GetDetailsOf(pidl2, col, &sd2)))
+                    {
+                        LPWSTR s1 = NULL, s2 = NULL;
+                        if (SUCCEEDED(hr = StrRetToStrW(&sd1.str, pidl1, &s1)) && SUCCEEDED(hr = StrRetToStrW(&sd2.str, pidl2, &s2)))
+                        {
+                            ret = StrCmpLogicalW(s1, s2);
+                            SHFree(s2);
+                            sd2.str.uType = STRRET_CSTR;
+                        }
+                        SHFree(s1);
+                    }
+                    FreeStrRet(sd2.str);
```

Comment:
`sd1` does not need a `FreeStrRet` ?

Reply:
I'll abstract away GetDetailsOf and the string handling in a helper function to make the code clear.